docs/src/main/asciidoc/class-loading-reference.adoc

```diff
@@ -0,0 +1,163 @@
+
+Deployment ClassLoader::
+
+This can load all application classes, it's parent is the Augment ClassLoader so it can also load all deployment classes.
+
+This ClassLoader is non-persistent, it will be re-created when the application is started, and is isolated. This ClassLoader
+is the context ClassLoader that is used when running the build steps. It is transformer safe so loading application classes
+will not prevent them from being transformed.
+
+Base Runtime ClassLoader::
+
+This loads all the runtime extension dependencies, as well as other user dependencies. It does not load the
+application root or any hot deployed code. This ClassLoader is persistent, even if the application restarts it will remain
+(which is why it cannot load application classes that may be hot deployed). It's parent is the base ClassLoader.
+
+This loads code that is not hot-reloadable, but it does support transformation (although once the class is loaded this
+transformation is no longer possible). This means that only transformers registered in the first application start
+will take effect, however as these transformers are expected to be idempotent this should not cause problems.
+
+This ClassLoader is isolated from the Augment and Deployment ClassLoaders. This means that it is not possible to set
+values in a static field in the deployment side, and expect to read it at runtime. This allows dev and test applications
+to behave more like a production application (production applications are isolated in that they run in a whole new JVM).
+
+This also means that the runtime version can be linked against a different set of dependencies, e.g. the hibernate
+version used at deployment time might want to include ByteBuddy, while the version used at runtime does not.
+
+Runtime Class Loader::
+
+This ClassLoader is used to load the application classes and other hot deployable resources. It's parent it the base runtime
+ClassLoader, and it is recreated when the application is restarted.
+
+
+== Isolated ClassLoaders
+
+The runtime ClassLoader is always isolated. This means that it will have it's own copies of almost every class from the
+resolved dependency list. The exception to this are:
+
+- JDK classes
+- Classes from artifacts that extensions have marked as parent first (more on this later).
+
+=== Parent First Dependencies
+
+There are some classes that should not be loaded in an isolated manner, but that should always be loaded by the system
+ClassLoader (or whatever ClassLoader is responsible for bootstrapping Quarkus). Most extensions do not need to worry about
+this, however there are a few cases where this is necessary:
+
+- Some logging related classes, as logging must be loaded by the system ClassLoader
+- Quarkus bootstrap itself
+
+If this is required it can be configured in the `quarkus-bootstrap-maven-plugin`. Note that if you
+mark a dependency as parent first then all of it's dependencies must also be parent first,
```

Comment:
```suggestion
mark a dependency as parent first then all of its dependencies must also be parent first,
```

Reply:
I think this needs to have more attention drawn to it. Extension authors would probably need to know more about this